.circleci/workflow-monitor.py

```diff
@@ -0,0 +1,63 @@
+#!/usr/bin/env python
+
+# Runs in the background on a manager instance to determine when it can be torn
+# down by polling the workflow's state using CircleCI's v2.0 restful api.
+#
+# Terminate instances if:
+#   - the workflow is successful (all jobs complete successfully)
+#   - the workflow is cancelled
+# Stop instances if:
+#   - the workflow has failed (all jobs have completed, but at least one has failed)
+#
+# Other states to consider: not_run, on_hold, unauthorized
+
+import httplib
+import time
+import sys
+import base64
+import json
+
+from common import terminate_workflow_instances, stop_workflow_instances
+
+# Time between HTTPS requests to circleci
+POLLING_INTERVAL_SECONDS = 60
+# Number of failed requests before stopping the instances
+QUERY_FAILURE_THRESHOLD = 10
+
+def main(workflow_id, circle_ci_token):
+    state = None
+    consecutive_failures = 0
+    auth_token = base64.b64encode(b"{}:", circle_ci_token)
+    headers = {'authorization': "Basic {}".format(auth_token)}
+
+    while True:
+        time.sleep(POLLING_INTERVAL_SECONDS)
+
+        conn = httplib.HTTPSConnection("circleci.com")
+        conn.request("GET", "/api/v2/workflow/{}".format(workflow_id), headers=headers)
+
+        res = conn.getresponse()
+
+        if res.status == httplib.OK:
+            consecutive_failures = 0
+            res_dict = json.load(res)
+            state = res_dict["status"]
+
+            print "Workflow {} status: {}".format(workflow_id, state)
+            if state == "success" or state == 'canceled':
+                terminate_workflow_instances(workflow_id)
+                exit(0)
+            elif state == "failed":
+                stop_workflow_instances(workflow_id)
+                exit(0)
+
```

Comment:
There are lots of states where we should do nothing, (e.g, failing --> we need to let the other jobs finish) maybe it's better if I explicitly codify that?

Reply:
I think it would be good to codify them.  To make the code less clunky, I'd probably use a dict of callables to map from the state to the action, like this:
```
# add an exit=None parameter to your terminate/stop_workflow_instances helpers
# and 'exit(exit) if exit' at the end of them.  How redundant is that?  Maybe call it exit_code?

term_and_exit_0 = lambda: terminate_workflow_instances(workflow_id, exit=0)
stop_and_exit_1 = lambda: terminate_workflow_instances(workflow_id, exit=0)
# my fabulous no-op lambda isn't needed anymore with the do_nothing_states factored
# out into a list that is checked after looking for an actionable state in wkflow_actions but here
# it is if you end up deciding to list the do_nothing_states in the wkflow_actions and get rid of the separate list
do_nothing = lambda: None

# setup closures to call based on workflow states where something needs to be done
wkflow_actions = {
   success:  term_and_exit_0
   canceled: term_and_exit_0
   failed: stop_and_exit_0
}

do_nothing_states = [
    'failing',
    # enumerate the remaining ones
]


action = wkflow_actions.get(state)
if action:
    action()
elif state in do_nothing_states:
    pass
else:
    print('this is where you should have an error condition where you complain about not knowing what to do')
    raise FreakOut
```

Reply:
Warning: I wrote the big block of code above only in the PR and didn't try to even run it through python .  Sharp edges.

Reply:
I'd still be worried about AWS adding some new state that this code doesn't handle but it isn't a big enough deal to postpone merging this.